drivers/sensor/adi/adxl372/adxl372_stream.c

```diff
@@ -0,0 +1,369 @@
+	struct rtio_iodev_sqe *current_sqe = data->sqe;
+	uint16_t fifo_samples = (((data->fifo_ent[0] & 0x3) << 8) | data->fifo_ent[1]);
+	size_t sample_set_size = adxl372_get_packet_size(cfg);
+
+	/* At least one sample set must remain in FIFO to encure that data
+	 * is not overwritten and stored out of order.
+	 */
+	if (fifo_samples > sample_set_size/2) {
+		fifo_samples -= sample_set_size/2;
+	} else {
+		LOG_ERR("fifo sample caunt error %d\n", fifo_samples);
+		gpio_pin_interrupt_configure_dt(&cfg->interrupt, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	uint16_t fifo_bytes = fifo_samples * 2/*sample size*/;
+
+	data->sqe = NULL;
+
+	/* Not inherently an underrun/overrun as we may have a buffer to fill next time */
+	if (current_sqe == NULL) {
+		LOG_ERR("No pending SQE");
+		gpio_pin_interrupt_configure_dt(&cfg->interrupt, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	const size_t min_read_size = sizeof(struct adxl372_fifo_data) + sample_set_size;
+	const size_t ideal_read_size = sizeof(struct adxl372_fifo_data) + fifo_bytes;
+
+	uint8_t *buf;
+	uint32_t buf_len;
+
+	if (rtio_sqe_rx_buf(current_sqe, min_read_size, ideal_read_size, &buf, &buf_len) != 0) {
+		LOG_ERR("Failed to get buffer");
+		rtio_iodev_sqe_err(current_sqe, -ENOMEM);
+		gpio_pin_interrupt_configure_dt(&cfg->interrupt, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+	LOG_DBG("Requesting buffer [%u, %u] got %u", (unsigned int)min_read_size,
+		(unsigned int)ideal_read_size, buf_len);
+
+	/* Read FIFO and call back to rtio with rtio_sqe completion */
+	struct adxl372_fifo_data hdr = {
+		.is_fifo = 1,
+		.timestamp = data->timestamp,
+		.int_status = data->status1,
+		.accel_odr = cfg->odr,
+		.sample_set_size = sample_set_size,
+	};
+
+	if ((cfg->fifo_config.fifo_format == ADXL372_X_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XY_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO)) {
+		hdr.has_x = 1;
+	}
+
+	if ((cfg->fifo_config.fifo_format == ADXL372_Y_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XY_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_YZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO))	{
+		hdr.has_y = 1;
+	}
+
+	if ((cfg->fifo_config.fifo_format == ADXL372_Z_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_YZ_FIFO) ||
+		(cfg->fifo_config.fifo_format == ADXL372_XYZ_FIFO))	{
+		hdr.has_z = 1;
+	}
+
+	uint32_t buf_avail = buf_len;
+
+	memcpy(buf, &hdr, sizeof(hdr));
+	buf_avail -= sizeof(hdr);
+
+	uint32_t read_len = MIN(fifo_bytes, buf_avail);
+	uint32_t pkts = read_len / sample_set_size;
+
+	read_len = pkts * sample_set_size;
+
+	((struct adxl372_fifo_data *)buf)->fifo_byte_count = read_len;
+
+	__ASSERT_NO_MSG(read_len % pkt_size == 0);
+
+	uint8_t *read_buf = buf + sizeof(hdr);
+
+	/* Flush completions */
+	struct rtio_cqe *cqe;
+
+	do {
```

Comment:
It's likely worth checking error codes here and bailing with rtio_iodev_sqe_err() on a bus error if one occurs.

Reply:
Can you give me more information? rtio_cqe_consume returns NULL when there is no CQE and that is cowered, and rtio_cqe_release is void func. 
Did you mean for this part: 
struct rtio_sqe *write_fifo_addr = rtio_sqe_acquire(data->rtio_ctx); ?

Reply:
When there is a completion queue event from the bus request, its result may be -errno, e.g.

```c
int res = 0;

do {
    cqe = rtio_cqe_consume(...);
    if (cqe != NULL) {
        if(cqe->result < 0 && res == 0) {
            /* handle spi/i2c bus error here */
            LOG_ERR("first bus error: %d", cqe->result);
            res = cqe->result;
        }
        rtio_cqe_release(cqe);
    }
} while (cqe != NULL);

/* Bail/cancel attempt to read sensor on any error */
if (res != 0) {
   rtio_iodev_sqe_err(...);
   return;
}
```

Reply:
That is a good idea. It is added in the latest code.